### PR TITLE
[ new ] Add implementation of functions generation

### PR DESCRIPTION
**Pull request by buzden**
Currently hedgehog has quite a lot of different useful standard generators and combinators of generators, but still lacks generation of functions. This is not an easy thing, however.

I propose a solution, that is, in fact, a blend of solutions from the Haskell's `hedgehog`, Haskell's `QuickCheck` and Scala's `scalacheck`. Generally, it provides two type of function generators, shrinkable one and non-shrinkable one. The first one is simpler in implementation, works faster and supports dependent types, at least to some extent. The second one is based on classic paper on shrinkable and showable functions, and indeed supports generating functions that can be shrunk well and shown nicely in a fail report. But it requires additional from a domain type, and (at least currently, and in the original research) does not support dependent types. So, both approaches are useful.

You can look at my playground that uses proposed generators [here](https://github.com/buzden/idris-playground/tree/master/functions-in-hedgehog). I added those experiments as a test for the hedgehog package, I think it's quite important to test. Since I tested failing behaviour, including output after a test fail and after test shrinking, I used `Test.Golden` library for that.

I suggested two new interfaces: one for types that can be in domain of generated functions, one for similar types for shrinkable functions. I provided several standard implementations. I also added a derivation configuration to be used with `elab-util`'s derivation mechanism, but I added those only for the first interface. I didn't manage to do this for the second one due to its complexity. Maybe, this is a task for future.

I need to say that I used interfaces here because being able to be in a domain of a generated function is really a property of a type, this differs from standard `Gen`s passing in Hedgehog.

### CODE REVIEW

**Comment by stefan-hoeck** [comment]
Thanks for adding this. This PR will take some time to digest but I'll try an be quick with reviewing it.

One big stumbling block I see: I'd rather not depend on `Test.Golden` if we can avoid it for reasons of maintainability.

File: src/Hedgehog/Internal/Function.idr
Comment (stefan-hoeck) [inline src/Hedgehog/Internal/Function.idr:84]: Could you please elaborate, why we use this implementation for all integral types with the exception of `Nat`? Would it not make more sense to injectively convert `Integer` to `Nat` and use `Cogen Nat` instead (see below)? All other integral types such as `BitsXY` and `IntXY` could go via `Integer`.

A possible injection (no loss of information) would map positive integers to even natural numbers and negative integers to odd ones: `if n >= 0 then cast (2 * n) else cast (2 * negate n + 1)`.

File: src/Hedgehog/Internal/Function.idr
Comment (stefan-hoeck) [inline src/Hedgehog/Internal/Function.idr:173]: Please use the following style for multi-line function declarations:

```idris
dargdepfun_ : 
     {0 b : a -> Type}
  -> {0 c : {0 x : a} -> b x -> Type}
  -> {auto _ : {0 x : a} -> Cogen (b x)}
  -> ({0 x : a} -> (y : b x) -> Gen (c y))
  -> Gen ({0 x : a} -> (y : b x) -> c y)
```

I try to use this style consistently throughout my codebase nowadays, but have not adjusted older code yet.

**Comment by buzden** [comment]
> One big stumbling block I see: I'd rather not depend on `Test.Golden` if we can avoid it for reasons of maintainability.

Oh, yes, I was afraid of it before doing so. But for testing purposes we need to check the output of failing and shrinking. I can see three variants:
- you put up with golden, since it anyway comes with the compiler, probably it's anyway well maintained;
- we remove tests that I added from this PR; this functionality shouldn't change a lot, maybe it's acceptable to check it occasionally;
- we change `check` and `recheck` to be able to accept alternative terminal which, say, uses `Writer` instead of `IO`; this will allow us to stay within hedgehog in tests too.

WDYT?

File: src/Hedgehog/Internal/Function.idr
Comment (stefan-hoeck) [inline src/Hedgehog/Internal/Function.idr:189]: This needs thorough documentation. Please add the link from above to the doc string but also explain in some details what this does and why it is needed.

Also, please consider giving this an identifier name instead of an operator. Otherwise, please explain why you think an operator is preferable in this case.

File: src/Hedgehog/Internal/Function.idr
Comment (stefan-hoeck) [inline src/Hedgehog/Internal/Function.idr:338]: Please add (minimal) documentation.

File: src/Hedgehog/Internal/Gen.idr
Comment (stefan-hoeck) [inline src/Hedgehog/Internal/Gen.idr:55]: Please add documentation. Also: Is it necessary to `public export` this?

File: src/Hedgehog/Internal/Seed.idr
Comment (stefan-hoeck) [inline src/Hedgehog/Internal/Seed.idr:141]: I don't understand what this means. Please document in such a way that users don't have to follow links to other libraries or papers.

File: src/Hedgehog/Internal/Function.idr
Comment (stefan-hoeck) [inline src/Hedgehog/Internal/Function.idr:138]: Please keep all lines (docs and code) at 80 or fewer characters unless strictly necessary otherwise.

**Review by stefan-hoeck — CHANGES_REQUESTED** [review]

**Comment by stefan-hoeck** [comment]
> > One big stumbling block I see: I'd rather not depend on `Test.Golden` if we can avoid it for reasons of maintainability.
> 
> Oh, yes, I was afraid of it before doing so. But for testing purposes we need to check the output of failing and shrinking. I can see three variants:
> 
> * you put up with golden, since it anyway comes with the compiler, probably it's anyway well maintained;
> * we remove tests that I added from this PR; this functionality shouldn't change a lot, maybe it's acceptable to check it occasionally;
> * we change `check` and `recheck` to be able to accept alternative terminal which, say, uses `Writer` instead of `IO`; this will allow us to stay within hedgehog in tests too.
> 
> WDYT?

The third option (staying in hedgehog) sounds preferable for me, but the necessary changes should go to a new PR.

**Comment by stefan-hoeck** [comment]
Sorry, seems I messed up some code locations in my review. Feel free to ask if something is not clear.

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr:173]: Sure, I just forgot. Since this was a new module, I used the style I'm used to without any additional thought, since there was no visual reference, as usual when I edit some existing module

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr:138]: Oh, really? I'll do so, if you insist, but do you really use such a narrow monitor? I thought that already a dozen of years standard length shifted to at least 120 characters.

Comment (stefan-hoeck) [reply src/Hedgehog/Internal/Function.idr:138]: I insist. It's the limit I use in all my projects. This is not about narrow monitors but about readability and using split screens a lot. Here are a couple of projects / organisations who adhere to 80 characters width:

[Jonathan Tibbel](https://github.com/tibbe/haskell-style-guide/blob/master/haskell-style.md)
[A CS Haskell course](https://www.cs.umd.edu/class/spring2022/cmsc488B/style.html)
[Another CS Haskell course; 78 characters only](http://courses.cms.caltech.edu/cs11/material/haskell/misc/haskell_style_guide.html)
[The official Scala style guide](https://docs.scala-lang.org/style/indentation.html)

File: src/Hedgehog/Internal/Function.idr
Comment (stefan-hoeck) [reply src/Hedgehog/Internal/Function.idr:173]: Looking through the hedgehog code I realize it needs some overhaul to be adjusted to my current coding style (it's one of my oldes Idris projects). I'll do that once this PR is shipped.

Comment (buzden) [reply src/Hedgehog/Internal/Function.idr]: Done

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr]: Done

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr]: Done. However, I don't understand why the convention is to reexport some stuff that is essential to a `derive` function work by a `Derive.*` module rather than by `Language.Reflection.Derive` (like, say, reexporting `Data.List1` and `Data.Vect`). I consider this to be a non-obvious implementation detail of the `derive` function, and making every `Derive.*` module to reexport the detail of implementation of `Language.Reflection.Derive` module seems strange to me.

File: src/Derive/Cogen.idr
Comment (stefan-hoeck) [inline src/Derive/Cogen.idr:3]: It should be sufficient to publicly export `Language.Reflection.Util` (see also `Derive.Prelude` in elab-util) plus any required hedgehog module(s).

**Review by stefan-hoeck — CHANGES_REQUESTED** [review]

File: src/Derive/Cogen.idr
Comment (buzden) [reply src/Derive/Cogen.idr]: Oh, indeed, changed

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr:86]: > Could you please elaborate, why we use this implementation for all integral types with the exception of `Nat`?

The reasons are that 1) `Nat` does not implement `Integral` 2) `String` does implement `Cast String Nat` in a way that disallows be to use `Cast a Nat => Cogen a` 3) I didn't know about your interface `ToInteger` 4) These instances are anyway needed for `ShrCogen`, thus I just reused them.

Now I see that `ToInteger` is, probably, the best interface for this, except for supporting `Double`s in non-optimal way. `Integral` in this case is again slightly better. To deal with it, we can require both `ToInteger` and `Integral`, WDYT?

File: src/Hedgehog/Internal/Seed.idr
Comment (buzden) [reply src/Hedgehog/Internal/Seed.idr]: I tried ;-)

File: src/Hedgehog/Internal/Gen.idr
Comment (buzden) [reply src/Hedgehog/Internal/Gen.idr]: I used `public export` only because all other functions around it upon the `Gen` type, like `toTree` and `mapGen` are also `public export`. Now I see that there is no used of this function, thus I removed it.

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr:388]: Done

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr:219]: I've added some, please take a look.

Speaking of an operator -- well, as almost always, operators are for nice look in signatures. Since this type is for specialised form of functions, it's nice to use an arrow for them. Moreover, this arrow is the same as in Haskell's Hedgehog and Quickcheck, thus, people from there will feel familiar with this operator.

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr:34]: Done

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr:76]: Done

File: src/Hedgehog/Internal/Function.idr
Comment (buzden) [reply src/Hedgehog/Internal/Function.idr]: Done

**Comment by buzden** [comment]
I tried to address every comment. Also I rebased upon #37 to address issue of dependence on `Test.Golden`, now all tests are done without external dependencies.

**Comment by buzden** [comment]
I've rebased this upon recently changed `main`

File: src/Hedgehog/Internal/Function.idr
Comment (stefan-hoeck) [reply src/Hedgehog/Internal/Function.idr:219]: This is really good now! Thanks for doing this thoroughly. It helps a lot with understanding what's going on and also with later maintaining the code.

File: src/Hedgehog/Internal/Function.idr
Comment (stefan-hoeck) [reply src/Hedgehog/Internal/Function.idr]: Personally, I'd prefer to not use a generic numeric implementation at all but define explicit implementations for all primitive types. My experience with generic implementations is - for the reasons you just listed - not a good one, unless the generic implementation truly holds for all cases. So, I'd do the following

```idris
Cogen Bits64 where
  perturb = variant

Cogen Bits64 where
  perturb n =
    if n >= 0
     then perturb (cast {to = Bits64} $ 2*n)
     else perturb (cast {to = Bits64} $ 2 * negate n + 1)

Cogen Nat where
  perturb = perturb . cast {to = Bits64}

Cogen Bits8 where
  perturb = perturb . cast {to = Bits64}

Cogen Int8 where
  perturn = perturb . cast {to = Integer}
```

For me, this is the last thing blocking this PR. After this, I am ready to merge.

Comment (buzden) [reply src/Hedgehog/Internal/Function.idr]: I did it this way. But during this change I thought that `Bits` interface can be the best of both worlds, at least `FiniteBits`.

Comment (buzden) [reply src/Hedgehog/Internal/Function.idr]: Oh, and if you wonder why `ShrCogen Nat` was done through `ShrCogen Integer`, and not separately, the reasons are 1) not to repeat similar code 2) I checked that current approach is must faster than special through `divmodNatNZ`

**Review by stefan-hoeck — APPROVED** [review]

**Pull request merged**
